presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java

```diff
@@ -669,7 +670,16 @@ private PlanWithProperties planAndEnforce(
 
             boolean preferenceSatisfied;
             if (pruneExtraHashSymbols) {
-                preferenceSatisfied = result.getHashSymbols().keySet().equals(requiredHashes.getHashes());
+                // Check for two conditions:
+                // (1) result.getHashSymbols() contains all requiredHashes;
+                // (2) result.getHashSymbols() is a superset of requiredHashes combined with preferredHashes (no pruning is needed)
+                Set<HashComputation> resultHashes = result.getHashSymbols().keySet();
+                preferenceSatisfied = resultHashes.containsAll(requiredHashes.getHashes())
```

Comment:
what was the issue here?

Reply:
@sopel39 

For plans where Join and RemoteExchange are separated by AssignUniqueId, this logic prevented the hash on join keys produced by RemoteExchange to be reused by the Join.

This plan (created by the new `PushUpAssignUniqueIdThroughRemoteExchange` rule)

```
- Join
   - AssignUniqueId
      - RemoteExchange
```

was changed to 

```
- Join
   - Project: (add hash on probe join key)
      - AssignUniqueId
          - Project: (drop hash on probe join key)
             - RemoteExchange
```

E.g.

This query 

```
SELECT name, (SELECT max(name) FROM region WHERE regionkey = nation.regionkey AND length(name) > length(nation.name)) FROM nation
```

was planned as

```
- Aggregate[<name, regionkey, unique>] => [name:varchar(25), regionkey:bigint, unique:bigint, max:varchar(25)]
        max := "max"("name_1")
    - LeftJoin[("regionkey" = "regionkey_0") AND ("length"("name_1") > "length"("name"))][$hashvalue_17, $hashvalue_18] => [name:varchar(25), regionkey:bigint, unique:bigint, name_1:varchar(25)]
            Distribution: PARTITIONED
        - Project[] => [name:varchar(25), regionkey:bigint, unique:bigint, $hashvalue_17:bigint]
                $hashvalue_17 := "combine_hash"(bigint '0', COALESCE("$operator$hash_code"("regionkey"), 0))
            - AssignUniqueId => [name:varchar(25), regionkey:bigint, unique:bigint]
                - Project[] => [name:varchar(25), regionkey:bigint]
                    - RemoteExchange[REPARTITION][$hashvalue] => name:varchar(25), regionkey:bigint, $hashvalue:bigint
                        - ScanProject[table = local:tpch:nation:sf0.01, originalConstraint = true] => [name:varchar(25), regionkey:bigint, $hashvalue_16:bigint]
```

instead of

```
- Aggregate[<name, regionkey, unique>] => [name:varchar(25), regionkey:bigint, unique:bigint, max:varchar(25)]
        max := "max"("name_1")
    - LeftJoin[("regionkey" = "regionkey_0") AND ("length"("name_1") > "length"("name"))][$hashvalue, $hashvalue_17] => [name:varchar(25), regionkey:bigint, unique:bigint, name_1:varchar(25)]
            Distribution: PARTITIONED
        - AssignUniqueId => [name:varchar(25), regionkey:bigint, $hashvalue:bigint, unique:bigint]
            - RemoteExchange[REPARTITION][$hashvalue] => name:varchar(25), regionkey:bigint, $hashvalue:bigint
                - ScanProject[table = local:tpch:nation:sf0.01, originalConstraint = true] => [name:varchar(25), regionkey:bigint, $hashvalue_16:bigint]
```